File: JavaScript/1358-Number of Substrings Containing All Three Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var numberOfSubstrings = function(s) {
    let charArr = new Array(3).fill(0);
    let left = 0, res = 0;
    const N = s.length;
    
    for (let i=0; i<s.length; i++){
        charArr[s[i].charCodeAt(0) - 'a'.charCodeAt(0)] += 1;
        while(charArr.every(x => x > 0)){
            res += N-i;
            charArr[s[left].charCodeAt(0) - 'a'.charCodeAt(0)] -= 1;
            left += 1;
        }
    }
    return res;
};

File: C++/1358-Number of Substrings Containing All Three Characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int N = s.size(), left = 0, res = 0;
        vector<int> cnt(3, 0);

        for (int i=0; i<N; i++) {
            cnt[s[i]-'a']++;
            while (cnt[0] >= 1 && cnt[1] >= 1 && cnt[2] >= 1) {
                res += N-i;
                cnt[s[left++]-'a']--;
            }
        }

        return res;
    }
};